config/file_config.go

```diff
@@ -92,336 +111,137 @@ type PrometheusMetricsConfig struct {
+	ActivationLevel           uint     `default:"90" validate:"gte=0,lte=100"`
+	DeactivationLevel         uint     `default:"75" validate:"gte=0,lte=100"`
+	StressSamplingRate        uint64   `default:"1000" validate:"gte=1"`
+	MinimumActivationDuration Duration `default:"10s"`
+	StartStressedDuration     Duration `default:"3s"`
+}
+
+// newConfig does the work of creating and loading the start of a config object
+// from the given arguments; if args is nil, it uses the command line arguments.
+// It's used by both the main init as well as the reload code.
+func newConfig(opts *CmdEnv) (*fileConfig, error) {
+	mainconf := &configContents{}
+	mainhash, err := readConfigInto(mainconf, opts.ConfigLocation, opts)
 	if err != nil {
 		return nil, err
 	}
 
-	r := viper.New()
-
-	r.SetDefault("Sampler", "DeterministicSampler")
-	r.SetDefault("SampleRate", 1)
-	r.SetDefault("DryRun", false)
-	r.SetDefault("DryRunFieldName", "refinery_kept")
-
-	r.SetConfigFile(rules)
-	err = r.ReadInConfig()
-
+	rulesconf := &rulesContents{}
+	ruleshash, err := readConfigInto(rulesconf, opts.RulesLocation, opts)
 	if err != nil {
 		return nil, err
 	}
 
-	fc := &fileConfig{
-		config:        c,
-		rules:         r,
-		conf:          &configContents{},
-		callbacks:     make([]func(), 0),
-		errorCallback: errorCallback,
-	}
-
-	err = fc.unmarshal()
-
-	if err != nil {
-		return nil, err
-	}
-
-	v := validator.New()
-	err = v.Struct(fc.conf)
-	if err != nil {
-		return nil, err
-	}
-
-	err = fc.validateGeneralConfigs()
-	if err != nil {
-		return nil, err
-	}
-
-	err = fc.validateSamplerConfigs()
-	if err != nil {
-		return nil, err
-	}
-
-	c.WatchConfig()
-	c.OnConfigChange(fc.onChange)
-
-	r.WatchConfig()
-	r.OnConfigChange(fc.onChange)
-
-	return fc, nil
-}
-
-func (f *fileConfig) onChange(in fsnotify.Event) {
-	v := validator.New()
-	err := v.Struct(f.conf)
-	if err != nil {
-		f.errorCallback(err)
-		return
-	}
-
-	err = f.validateGeneralConfigs()
-	if err != nil {
-		f.errorCallback(err)
-		return
-	}
-
-	err = f.validateSamplerConfigs()
-	if err != nil {
-		f.errorCallback(err)
-		return
+	cfg := &fileConfig{
+		mainConfig:  mainconf,
+		mainHash:    mainhash,
+		rulesConfig: rulesconf,
+		rulesHash:   ruleshash,
+		opts:        opts,
 	}
 
-	f.unmarshal()
-
-	for _, c := range f.callbacks {
-		c()
-	}
+	return cfg, nil
 }
 
-func (f *fileConfig) unmarshal() error {
-	f.mux.Lock()
-	defer f.mux.Unlock()
-	err := f.config.Unmarshal(f.conf)
-
+// NewConfig creates a new Config object from the given arguments; if args is
+// nil, it uses the command line arguments
+func NewConfig(opts *CmdEnv, errorCallback func(error)) (Config, error) {
+	cfg, err := newConfig(opts)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	err = f.rules.Unmarshal(f.conf)
+	cfg.callbacks = make([]func(), 0)
+	cfg.errorCallback = errorCallback
 
-	if err != nil {
-		return err
-	}
+	go cfg.monitor()
```

Comment:
I don't think so, because it would imply that hot reloading would become an option, and also means that a lot of tests would have to make explicit calls to Start(). I don't personally feel that cost is justified by the symmetry.